### PR TITLE
Missing outputs should be recorded as test errors

### PR DESCRIPTION
**Pull request by bernt-matthias**
from `planemo` the `verify_tool` function is called from [within an try-except block](https://github.com/galaxyproject/planemo/blob/1aa3eb05a97ad20c0be6f6560ab5cec090e76612/planemo/engine/galaxy.py#L109) which silently catches any exception.

Thus any exception raised from within verify_tool will not be detected, i.e. the assertion needs to be moved into `_verify_outputs` (which also seems to make sense by name) in order to make verify_tool record the problem properly.

This only affects tool tests. Guess we do not want to actually forbid tool runs with no outputs.. Even if I don't know what they are used for. 
## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/tool_util/verify/interactor.py [hunk 1]
```diff
@@ -1511,6 +1509,10 @@ def register_exception(e: Exception):
 
     job_stdio = galaxy_interactor.get_job_stdio(job["id"])
 
+    if not (data_list or data_collection_list):
+        error = AssertionError("Tool produced no output datasets or collections")
+        register_exception(error)
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/interactor.py]:
```suggestion
    assert data_list or data_collection_list, "Tool produced no output datasets or collections"
```

File: test/functional/tools/output_filter.xml [hunk 2]
```diff
@@ -37,134 +36,113 @@ echo 'test' > p2.reverse
...
-            </output>
-            <output name="out_3">
-                <assert_contents>
-                    <has_line line="test" />
+                    <has_line line="1" />
                 </assert_contents>
             </output>
         </test>
-        <test expect_num_outputs="1">
+        <!-- tool runs with no outputs should fail -->
```
Comment (mvdbeek) [inline test/functional/tools/output_filter.xml:62]: I don't see why that should be the case ?

File: test/functional/tools/output_filter.xml [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
-echo 'test' > p1.forward &&
-echo 'test' > p1.reverse &&
-echo 'test' > p2.forward &&
-echo 'test' > p2.reverse 
+echo '1' > 1 &&
+echo '2' > 2 &&
+echo '3' > 3 &&
+echo '4' > 4 &&
+echo '5' > 5 &&
+echo 'p1.forward' > p1.forward &&
```
Comment (mvdbeek) [inline test/functional/tools/output_filter.xml:9]: Is that just refactoring or does this prove something in the tests ?

**Comment by mvdbeek** [comment]
A minor remark on the PR title: In general it is a good idea for the title to document what the PR does, instead of how you did something (which is pretty clear from the code anyway). So in this case I'd say something like "Make sure that missing outputs are recorded as test errors" helps a lot, both for the docs and the reviewer.

File: test/functional/tools/output_filter.xml (hunk 2 above)
Comment (bernt-matthias) [reply test/functional/tools/output_filter.xml:62]: This was not precise enough. Maybe the comment should say: 

```suggestion
        <!-- a test that does not produce any outputs violates the silent assumption that a tool run should have an output, hence expect_test_failure -->
```

Background if planemo finds a test that does not produce outputs it will skip all output checks and just log the assertion to the terminal. So the user will very likely miss this (in particular in CI). 

We might also remove the assertion completely.

Are there tools that do not produce outputs?

Comment (bernt-matthias) [reply test/functional/tools/output_filter.xml:62]: So .. the tool itself does not fail (I used `expect_test_failure` and not `expect_failure`).

File: lib/galaxy/tool_util/verify/interactor.py (hunk 1 above)
Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/interactor.py]: We need to (at least it's the idea of the PR) record the assertion error to check for `expect_test_failure` at the end.

File: test/functional/tools/output_filter.xml (hunk 2 above)
Comment (mvdbeek) [reply test/functional/tools/output_filter.xml:62]: > a test that does not produce any outputs violates the silent assumption that a tool run should have an output, hence expect_test_failure

what is this silent assumption ? an export tool doesn't need to create outputs

File: lib/galaxy/tool_util/verify/interactor.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/verify/interactor.py]: I don't understand what you're saying. `expect_test_failure` is set by the test definition. What are you trying to fix ?

Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/interactor.py]: We need to record the exception (i.e. call `register_exception` which stores in `found_exceptions`) in order to check for `expect_test_failure` https://github.com/galaxyproject/galaxy/blob/288a6c989a37be04a357191aeae9711f3b714806/lib/galaxy/tool_util/verify/interactor.py#L1625

If we just raise the Assertion then we have the same behavior as before. If we register it, we can react to this.

File: test/functional/tools/output_filter.xml (hunk 2 above)
Comment (bernt-matthias) [reply test/functional/tools/output_filter.xml:62]: Maybe implicit assumption is better. Anyway the export tool is a good example that the current change is not good.

The bottom line is the following. Currently, no check of the outputs is done if a tool does not produce any output but an assertion is raised. This assertion only appears in the logs and (for instance) planemo users don't see it if they are not paying attention to the logs (which probably only a few people do).

In other words, tool tests are successful if the tool test does not produce any output regardless of if there are any test output assumptions.

Or even shorter: the `AssertionError` short circuits all test outputs assumptions. 

... Maybe just remove the assertion?

File: lib/galaxy/tool_util/verify/interactor.py [hunk 4]
```diff
@@ -1511,6 +1509,10 @@ def register_exception(e: Exception):
...
+
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/verify/interactor.py]:
```suggestion
```

File: test/functional/tools/output_filter.xml (hunk 3 above)
Comment (bernt-matthias) [reply test/functional/tools/output_filter.xml:9]: Just refactoring. Wanted to be a bit more sure/explicit in the tests about which file is which.

**Comment by jdavcs** [comment]
@bernt-matthias I'm bumping this to 24.1. Feel free to re-milestone.

File: lib/galaxy/tool_util/verify/interactor.py [hunk 5]
```diff
@@ -1503,6 +1500,9 @@ def register_exception(e: Exception):
         if testdef.outputs:
             raise Exception("Cannot specify outputs in a test expecting failure.")
 
+    if not data_list and not data_collection_list:
+        Exception("Job did not produce any outputs")
+
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/verify/interactor.py]:
```suggestion
```

File: lib/galaxy/tool_util/verify/interactor.py [hunk 6]
```diff
@@ -1516,6 +1513,8 @@ def register_exception(e: Exception):
 
     job_stdio = galaxy_interactor.get_job_stdio(job["id"])
 
+    assert data_list or data_collection_list, "Tool produced no output datasets or collections"
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/interactor.py]: That's only a bug if we expected an output. Pure export tools don't need to produce outputs.

File: test/functional/tools/output_filter.xml [hunk 7]
```diff
@@ -37,134 +36,113 @@ echo 'test' > p2.reverse
...
+        <!-- tool runs with no outputs should fail -->
+        <test expect_num_outputs="0" expect_test_failure="true">
             <param name="produce_out_1" value="false" />
             <param name="filter_text_1" value="not_foo_or_bar" />
-            <output name="out_3">
-                <assert_contents>
-                    <has_line line="test" />
-                </assert_contents>
-            </output>
+            <assert_stdout>
```
Comment (bernt-matthias) [inline test/functional/tools/output_filter.xml:71]:
```suggestion
            <output name="out_3">
                <assert_contents>
                    <has_line line="test" />
                </assert_contents>
            </output>
            <assert_stdout>
                <has_n_lines n="0"/>
            </assert_stdout>
```

File: lib/galaxy/tool_util/verify/interactor.py (hunk 6 above)
Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/interactor.py]: I see.

File: test/functional/tools/output_filter.xml [hunk 8]
```diff
@@ -37,134 +36,120 @@ echo 'test' > p2.reverse
...
             <output name="out_3">
                 <assert_contents>
                     <has_line line="test" />
                 </assert_contents>
             </output>
+            <assert_stdout>
+                <has_n_lines n="0"/>
+            </assert_stdout>
+                <has_n_lines n="0"/>
+            </assert_stdout>
```
Comment (bernt-matthias) [inline test/functional/tools/output_filter.xml]:
```suggestion
```

File: lib/galaxy/tool_util/verify/interactor.py [hunk 9]
```diff
@@ -1547,23 +1544,29 @@ def register_exception(e: Exception):
...
-            )
-        except Exception as e:
-            register_exception(e)
+            try:
+                if hasattr(data_list, "values"):
+                    output_data = list(data_list.values())[output_index]
+                else:
+                    output_data = data_list[len(data_list) - len(testdef.outputs) + output_index]
+            except IndexError:
+                pass
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/interactor.py]: That seems weird, can you add a comment here ? I think that's also why the framework test failed.

Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/interactor.py]: You are right. How about https://github.com/galaxyproject/galaxy/pull/16094/commits/6ee5b09f95b2fbd06dc59675d72d1dd68dd9431d? Hope this makes it clearer. The comment would have been more or less the content of the exception we are recording here.

File: lib/galaxy/tool_util/verify/interactor.py [hunk 10]
```diff
@@ -1541,29 +1538,34 @@ def register_exception(e: Exception):
...
-        except Exception as e:
-            register_exception(e)
+            try:
+                if hasattr(data_list, "values"):
+                    output_data = list(data_list.values())[output_index]
+                else:
+                    output_data = data_list[len(data_list) - len(testdef.outputs) + output_index]
+            except IndexError:
+                error = AssertionError(f"Tool did not produce an output with name '{name}' (or at index {output_index})")
+                register_exception(error)
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/interactor.py:1557]: `output_data` is still undefined below. maybe set it to None ?

Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/interactor.py:1557]: This is also in https://github.com/galaxyproject/galaxy/pull/16094/commits/6ee5b09f95b2fbd06dc59675d72d1dd68dd9431d Just before the `try`. Or do I miss something?

Comment (mvdbeek) [reply lib/galaxy/tool_util/verify/interactor.py:1557]: yes, didn't see that.

**Review by mvdbeek — APPROVED** [review]

**Comment by bernt-matthias** [comment]
Should we target 24.0?

**Comment by jdavcs** [comment]
@bernt-matthias I'm merging this into dev - I don't think it's a bug (as per discussion above)? I've also edited the title - please feel free to change it if it's not a good summary of the PR. As per this [comment](https://github.com/galaxyproject/galaxy/pull/16094#issuecomment-1658680559), we'd have to manually edit this for the 24.1 release notes.

**Comment by mvdbeek** [comment]
It is definitely a bug, 24.0 is a good idea

**Comment by bernt-matthias** [comment]
https://github.com/galaxyproject/galaxy/pull/17873

**Comment by bernt-matthias** [comment]
Not sure how to retarget without adding loads of commits?

**Comment by jdavcs** [comment]
> Not sure how to retarget without adding loads of commits?

You should be able to edit the target: click the 'edit' button at the top of the page - it'll let you edit the title AND the base branch: you'll see a drop down list. Select '24.0 release' - it'll ask you to confirm and that's it. There may or may not be conflicts - is that what you were referring to?

**Comment by bernt-matthias** [comment]
If I do this I get 215 commits.

**Comment by jdavcs** [comment]
> If I do this I get 215 commits.

Oh no! Would you like to do #17873 *instead* of this (adding the extra commit to it)? i.e., merge it into 24.0 and then, if there are conflicts, manually forward-merge it into dev?

**Comment by bernt-matthias** [comment]
All this is now in https://github.com/galaxyproject/galaxy/pull/17874 (had to create a new PR since I force pushed before reopening .. and apparently one cannot reopen in this case).
